Exception when instantiating ActionServer as part of ROS package tests

## Bug report

**Required Info:**

- Operating System:
  - Ubuntu 22.04
- Installation type:
  - Binaries
- Version or commit hash:
  - Latest binaries of Humble/Iron/Rolling
- DDS implementation:
  - Fast DDS or Cyclone DDS (doesn't make a difference)
- Client library (if applicable):
  - rclpy

#### Steps to reproduce issue
I have a test that is failing as part of this PR (https://github.com/RobotWebTools/rosbridge_suite/pull/886) when I try to instantiate an action server, noting that:

* This only happens on Iron and Rolling, but not on Humble
* This only happens when I run using `colcon test` (which in turn uses `ament_add_pytest_test` CMake capability), but not if I run it directly using Python

Can't seem to figure out why this would be the case, so figured I'd elevate it here? Thank you!

The code:
```python
from example_interfaces.action import Fibonacci

class MyClass:
    def __init__(self, node):
        self.action_server = ActionServer(
            node,  # Just an rclpy.Node
            Fibonacci,
            "/fibonacci_action_1",
            self.execute_callback,
        )
```

The error:
```
          self.action_server = ActionServer(
      /opt/ros/rolling/lib/python3.10/site-packages/rclpy/action/server.py:264: in __init__
          self._logger = self._node.get_logger().get_child('action_server')
      _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
      
      self = <rclpy.impl.rcutils_logger.RcutilsLogger object at 0x7f730bfe9270>
      name = 'action_server'
      
          def get_child(self, name):
              if not name:
                  raise ValueError('Child logger name must not be empty.')
          
              if self.name:
                  # Prepend the name of this logger
      >           _rclpy.rclpy_logging_rosout_add_sublogger(self.name, name)
      E           RuntimeError: Failed to call rcl_logging_rosout_add_sublogger
```

#### Expected behavior
Should not raise an exception.

#### Actual behavior
Raises an exception.

--- a/rcl/include/rcl/logging_rosout.h
+++ b/rcl/include/rcl/logging_rosout.h
@@ -213,7 +213,7 @@ rcl_logging_rosout_output_handler(
  * \param[in] sublogger_name a sublogger name
  * \return #RCL_RET_OK if the subordinate logger was created successfully, or
  * \return #RCL_RET_INVALID_ARGUMENT if any arguments are invalid, or
- * \return #RCL_RET_SUBLOGGER_ALREADY_EXIST if the subordinate logger already exists, or
+ * \return #RCL_RET_NOT_FOUND if the parent logger not exists, or
  * \return #RCL_RET_BAD_ALLOC if allocating memory failed, or
  * \return #RCL_RET_ERROR if an unspecified error occurs.
  */
@@ -242,6 +242,7 @@ rcl_logging_rosout_add_sublogger(
  * \param[in] sublogger_name a sublogger name
  * \return #RCL_RET_OK if the subordinate logger was finalized successfully, or
  * \return #RCL_RET_INVALID_ARGUMENT if any arguments are invalid, or
+ * \return #RCL_RET_NOT_FOUND if the sublogger not exists, or
  * \return #RCL_RET_BAD_ALLOC if allocating memory failed, or
  * \return #RCL_RET_ERROR if an unspecified error occurs.
  */

--- a/rcl/include/rcl/logging_rosout.h
+++ b/rcl/include/rcl/logging_rosout.h
@@ -213,7 +213,7 @@ rcl_logging_rosout_output_handler(
  * \param[in] sublogger_name a sublogger name
  * \return #RCL_RET_OK if the subordinate logger was created successfully, or
  * \return #RCL_RET_INVALID_ARGUMENT if any arguments are invalid, or
- * \return #RCL_RET_NOT_FOUND if the parent logger not exists, or
+ * \return #RCL_RET_NOT_FOUND if the parent logger does not exist, or
  * \return #RCL_RET_BAD_ALLOC if allocating memory failed, or
  * \return #RCL_RET_ERROR if an unspecified error occurs.
  */
@@ -242,7 +242,7 @@ rcl_logging_rosout_add_sublogger(
  * \param[in] sublogger_name a sublogger name
  * \return #RCL_RET_OK if the subordinate logger was finalized successfully, or
  * \return #RCL_RET_INVALID_ARGUMENT if any arguments are invalid, or
- * \return #RCL_RET_NOT_FOUND if the sublogger not exists, or
+ * \return #RCL_RET_NOT_FOUND if the sublogger does not exist, or
  * \return #RCL_RET_BAD_ALLOC if allocating memory failed, or
  * \return #RCL_RET_ERROR if an unspecified error occurs.
  */

--- a/rcl/include/rcl/types.h
+++ b/rcl/include/rcl/types.h
@@ -51,6 +51,8 @@ typedef rmw_ret_t rcl_ret_t;
 #define RCL_RET_UNKNOWN_SUBSTITUTION 105
 /// rcl_shutdown() already called return code.
 #define RCL_RET_ALREADY_SHUTDOWN 106
+/// Resource not found
+#define RCL_RET_NOT_FOUND 107
 
 // rcl node specific ret codes in 2XX
 /// Invalid rcl_node_t given return code.

--- a/rcl/src/rcl/common.c
+++ b/rcl/src/rcl/common.c
@@ -57,6 +57,8 @@ rcl_convert_rcutils_ret_to_rcl_ret(rcutils_ret_t rcutils_ret)
       return RCL_RET_INVALID_ARGUMENT;
     case RCUTILS_RET_NOT_INITIALIZED:
       return RCL_RET_NOT_INIT;
+    case RCUTILS_RET_NOT_FOUND:
+      return RCL_RET_NOT_FOUND;
     default:
       return RCL_RET_ERROR;
   }

--- a/rcl/src/rcl/logging_rosout.c
+++ b/rcl/src/rcl/logging_rosout.c
@@ -442,11 +442,11 @@ rcl_logging_rosout_add_sublogger(
   RCL_CHECK_ARGUMENT_FOR_NULL(logger_name, RCL_RET_INVALID_ARGUMENT);
   RCL_CHECK_ARGUMENT_FOR_NULL(sublogger_name, RCL_RET_INVALID_ARGUMENT);
   rcutils_ret_t rcutils_ret = rcutils_hash_map_get(&__logger_map, &logger_name, &entry);
-  if (RCL_RET_OK != (status = rcl_convert_rcutils_ret_to_rcl_ret(rcutils_ret))) {
-    if (RCL_RET_NOT_FOUND == status) {
+  if (RCUTILS_RET_OK != rcutils_ret) {
+    if (RCUTILS_RET_NOT_FOUND == rcutils_ret) {
       RCL_SET_ERROR_MSG_WITH_FORMAT_STRING("Failed to get logger entry for '%s'.", logger_name);
     }
-    return status;
+    return rcl_convert_rcutils_ret_to_rcl_ret(rcutils_ret);
   }
 
   status =

--- a/rcl/src/rcl/logging_rosout.c
+++ b/rcl/src/rcl/logging_rosout.c
@@ -442,7 +442,7 @@ rcl_logging_rosout_add_sublogger(
   RCL_CHECK_ARGUMENT_FOR_NULL(sublogger_name, RCL_RET_INVALID_ARGUMENT);
   rcutils_ret_t rcutils_ret = rcutils_hash_map_get(&__logger_map, &logger_name, &entry);
   if (RCUTILS_RET_NOT_FOUND == rcutils_ret) {
-    return RCL_RET_OK;
+    return RCL_RET_NOT_FOUND;
   }
 
   status =
@@ -528,7 +528,7 @@ rcl_logging_rosout_remove_sublogger(
   }
 
   if (!rcutils_hash_map_key_exists(&__logger_map, &full_sublogger_name)) {
-    status = RCL_RET_OK;
+    status = RCL_RET_NOT_FOUND;
     goto cleanup;
   }
 

--- a/rcl/src/rcl/logging_rosout.c
+++ b/rcl/src/rcl/logging_rosout.c
@@ -438,18 +438,18 @@ rcl_logging_rosout_add_sublogger(
   rosout_map_entry_t entry;
   rosout_sublogger_entry_t sublogger_entry;
 
+  RCL_CHECK_ARGUMENT_FOR_NULL(logger_name, RCL_RET_INVALID_ARGUMENT);
+  RCL_CHECK_ARGUMENT_FOR_NULL(sublogger_name, RCL_RET_INVALID_ARGUMENT);
+  rcutils_ret_t rcutils_ret = rcutils_hash_map_get(&__logger_map, &logger_name, &entry);
+  if (RCUTILS_RET_NOT_FOUND == rcutils_ret) {
+    return RCL_RET_OK;
+  }
+
   status =
     _rcl_logging_rosout_get_full_sublogger_name(logger_name, sublogger_name, &full_sublogger_name);
   if (RCL_RET_OK != status) {
     // Error already set
     return status;
-  }
-
-  rcutils_ret_t rcutils_ret = rcutils_hash_map_get(&__logger_map, &logger_name, &entry);
-  if (RCUTILS_RET_OK != rcutils_ret) {
-    RCL_SET_ERROR_MSG_WITH_FORMAT_STRING("The entry of logger '%s' not exist.", logger_name);
-    status = RCL_RET_ERROR;
-    goto cleanup;
   }
 
   if (rcutils_hash_map_key_exists(&__logger_map, &full_sublogger_name)) {
@@ -517,6 +517,8 @@ rcl_logging_rosout_remove_sublogger(
   rcl_ret_t status = RCL_RET_OK;
   char * full_sublogger_name = NULL;
   rosout_sublogger_entry_t sublogger_entry;
+  RCL_CHECK_ARGUMENT_FOR_NULL(logger_name, RCL_RET_INVALID_ARGUMENT);
+  RCL_CHECK_ARGUMENT_FOR_NULL(sublogger_name, RCL_RET_INVALID_ARGUMENT);
 
   status =
     _rcl_logging_rosout_get_full_sublogger_name(logger_name, sublogger_name, &full_sublogger_name);
@@ -525,13 +527,12 @@ rcl_logging_rosout_remove_sublogger(
     return status;
   }
 
-  // remove the entry from the map
   if (!rcutils_hash_map_key_exists(&__logger_map, &full_sublogger_name)) {
-    RCL_SET_ERROR_MSG_WITH_FORMAT_STRING("Sub-logger '%s' not exist.", full_sublogger_name);
-    status = RCL_RET_ERROR;
+    status = RCL_RET_OK;
     goto cleanup;
   }
 
+  // remove the entry from the map
   status = rcl_ret_from_rcutils_ret(
     rcutils_hash_map_get(&__sublogger_map, &full_sublogger_name, &sublogger_entry));
   if (RCL_RET_OK != status) {

--- a/rcl/src/rcl/logging_rosout.c
+++ b/rcl/src/rcl/logging_rosout.c
@@ -15,6 +15,7 @@
 #include "rcl/logging_rosout.h"
 
 #include "rcl/allocator.h"
+#include "rcl/common.h"
 #include "rcl/error_handling.h"
 #include "rcl/node.h"
 #include "rcl/publisher.h"
@@ -441,8 +442,11 @@ rcl_logging_rosout_add_sublogger(
   RCL_CHECK_ARGUMENT_FOR_NULL(logger_name, RCL_RET_INVALID_ARGUMENT);
   RCL_CHECK_ARGUMENT_FOR_NULL(sublogger_name, RCL_RET_INVALID_ARGUMENT);
   rcutils_ret_t rcutils_ret = rcutils_hash_map_get(&__logger_map, &logger_name, &entry);
-  if (RCUTILS_RET_NOT_FOUND == rcutils_ret) {
-    return RCL_RET_NOT_FOUND;
+  if (RCL_RET_OK != (status = rcl_convert_rcutils_ret_to_rcl_ret(rcutils_ret))) {
+    if (RCL_RET_NOT_FOUND == status) {
+      RCL_SET_ERROR_MSG_WITH_FORMAT_STRING("Failed to get logger entry for '%s'.", logger_name);
+    }
+    return status;
   }
 
   status =
@@ -529,6 +533,7 @@ rcl_logging_rosout_remove_sublogger(
 
   if (!rcutils_hash_map_key_exists(&__logger_map, &full_sublogger_name)) {
     status = RCL_RET_NOT_FOUND;
+    RCL_SET_ERROR_MSG_WITH_FORMAT_STRING("Logger for '%s' not found.", full_sublogger_name);
     goto cleanup;
   }
 

--- a/rcl/test/rcl/test_logging_rosout.cpp
+++ b/rcl/test/rcl/test_logging_rosout.cpp
@@ -380,8 +380,8 @@ TEST_F(
   rcl_reset_error();
   EXPECT_EQ(RCL_RET_INVALID_ARGUMENT, rcl_logging_rosout_add_sublogger(logger_name, ""));
   rcl_reset_error();
-  EXPECT_EQ(RCL_RET_OK, rcl_logging_rosout_add_sublogger("", "child"));
-  EXPECT_EQ(RCL_RET_OK, rcl_logging_rosout_add_sublogger("no_exist", "child"));
+  EXPECT_EQ(RCL_RET_NOT_FOUND, rcl_logging_rosout_add_sublogger("", "child"));
+  EXPECT_EQ(RCL_RET_NOT_FOUND, rcl_logging_rosout_add_sublogger("no_exist", "child"));
 
   EXPECT_EQ(RCL_RET_INVALID_ARGUMENT, rcl_logging_rosout_remove_sublogger(nullptr, nullptr));
   rcl_reset_error();
@@ -405,7 +405,7 @@ TEST_F(
 
   EXPECT_EQ(RCL_RET_OK, rcl_logging_rosout_add_sublogger(logger_name, "child2"));
   EXPECT_EQ(RCL_RET_OK, rcl_logging_rosout_remove_sublogger(logger_name, "child2"));
-  EXPECT_EQ(RCL_RET_OK, rcl_logging_rosout_remove_sublogger(logger_name, "child2"));
+  EXPECT_EQ(RCL_RET_NOT_FOUND, rcl_logging_rosout_remove_sublogger(logger_name, "child2"));
   rcl_reset_error();
 }
 

--- a/rcl/test/rcl/test_logging_rosout.cpp
+++ b/rcl/test/rcl/test_logging_rosout.cpp
@@ -381,7 +381,9 @@ TEST_F(
   EXPECT_EQ(RCL_RET_INVALID_ARGUMENT, rcl_logging_rosout_add_sublogger(logger_name, ""));
   rcl_reset_error();
   EXPECT_EQ(RCL_RET_NOT_FOUND, rcl_logging_rosout_add_sublogger("", "child"));
+  rcl_reset_error();
   EXPECT_EQ(RCL_RET_NOT_FOUND, rcl_logging_rosout_add_sublogger("no_exist", "child"));
+  rcl_reset_error();
 
   EXPECT_EQ(RCL_RET_INVALID_ARGUMENT, rcl_logging_rosout_remove_sublogger(nullptr, nullptr));
   rcl_reset_error();

--- a/rcl/test/rcl/test_logging_rosout.cpp
+++ b/rcl/test/rcl/test_logging_rosout.cpp
@@ -380,10 +380,8 @@ TEST_F(
   rcl_reset_error();
   EXPECT_EQ(RCL_RET_INVALID_ARGUMENT, rcl_logging_rosout_add_sublogger(logger_name, ""));
   rcl_reset_error();
-  EXPECT_EQ(RCL_RET_INVALID_ARGUMENT, rcl_logging_rosout_add_sublogger("", "child"));
-  rcl_reset_error();
-  EXPECT_EQ(RCL_RET_ERROR, rcl_logging_rosout_add_sublogger("no_exist", "child"));
-  rcl_reset_error();
+  EXPECT_EQ(RCL_RET_OK, rcl_logging_rosout_add_sublogger("", "child"));
+  EXPECT_EQ(RCL_RET_OK, rcl_logging_rosout_add_sublogger("no_exist", "child"));
 
   EXPECT_EQ(RCL_RET_INVALID_ARGUMENT, rcl_logging_rosout_remove_sublogger(nullptr, nullptr));
   rcl_reset_error();
@@ -407,7 +405,7 @@ TEST_F(
 
   EXPECT_EQ(RCL_RET_OK, rcl_logging_rosout_add_sublogger(logger_name, "child2"));
   EXPECT_EQ(RCL_RET_OK, rcl_logging_rosout_remove_sublogger(logger_name, "child2"));
-  EXPECT_EQ(RCL_RET_ERROR, rcl_logging_rosout_remove_sublogger(logger_name, "child2"));
+  EXPECT_EQ(RCL_RET_OK, rcl_logging_rosout_remove_sublogger(logger_name, "child2"));
   rcl_reset_error();
 }
 

